"Write a Post" view redesign

ludwiczakpawel opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

This PR introduces new styling for "Write a Post" view.

## Related Tickets & Documents

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

![image](https://user-images.githubusercontent.com/108287/80370961-65eeb680-8891-11ea-8bbf-0dc33a90ada7.png)


## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [x] no documentation needed

nickytonline commented:
The PR is quite large @ludwiczakpawel. Is there any way to break this down into smaller PRs, or is it an all in or nothing type of situation?

I know this is a work in progress, so you may already be aware of some of these issues. I've pulled this locally and I'm getting some errors.

Unable to upload an image for the content

![image](https://user-images.githubusercontent.com/833231/80627354-c4e13680-8a1d-11ea-87d9-234391348512.png)

Saving as a draft I get the following

![image](https://user-images.githubusercontent.com/833231/80627408-da566080-8a1d-11ea-9598-fa91a0e077b9.png)

Probably not related but putting it here anyways. The onboarding styling appears to be all messed up

![image](https://user-images.githubusercontent.com/833231/80627518-112c7680-8a1e-11ea-9ebb-5e945d2accd3.png)

On mobile, buttons don't end up being the same size due to the content

![image](https://user-images.githubusercontent.com/833231/80627665-45079c00-8a1e-11ea-819e-1d993a74bec1.png)

Options for Canonical URL and Series only partially visible on mobile

![image](https://user-images.githubusercontent.com/833231/80627875-8d26be80-8a1e-11ea-80df-081a4810fa5e.png)

The change the cover button ends up becoming a square due to the length of the content. Maybe we could simply make the button text "Change"?

![image](https://user-images.githubusercontent.com/833231/80627999-b9423f80-8a1e-11ea-88e2-16937994373b.png)

Placeholder text for the content of a post is cut off on mobile

![image](https://user-images.githubusercontent.com/833231/80628062-cfe89680-8a1e-11ea-871a-09aa01f19c44.png)

This is most likely due to the fact the text area is larger than the width on mobile

![image](https://user-images.githubusercontent.com/833231/80628163-ef7fbf00-8a1e-11ea-92b9-45819ba829dc.png)

Error messages are also cut off on Mobile

![image](https://user-images.githubusercontent.com/833231/80628214-01f9f880-8a1f-11ea-9a12-a5c607cc7a1f.png)

ludwiczakpawel commented:
Thanks @nickytonline ! yeap, i'm aware of some of these issues but not all so thanks for reporting!! I'll take care of these! I'll also try to split SOME things out of this PR to make it at least a tiny bit lighter :) I'll follow up with some questions re your issues tomorrow probably.

benhalpern commented:
This is looking really fabulous. At a glance I don't see any gotchas or issues with the code, let me know if I can be helpful in talking through anything on the TODOs. I'll keep circling back as it gets closer.

Definitely worth making sure the tests are of good quality here to establish a good foundation.

ludwiczakpawel commented:
@nickytonline to follow up on some of the issues:

> Unable to upload an image for the content
&
> Saving as a draft I get the following

Can you check whether you can do these two actions on **master** locally? I'm trying to figure out if it's because of local env or is it because of some of my changes. Asking because I can easily upload images for example..

---

> onboarding styling appears to be all messed up

This one should be already fixed BUT I have created a separate PR that addresses only modal-related issues ( #7603 ) so this one can be a bit lighter.

---

Responsiveness should now be a bit better :)

Ridhwana commented on app/assets/stylesheets/config/_import.scss:
```diff
@@ -48,6 +48,8 @@ $colors: (
 
 $header-height: 56px;
 $site-width: 1280px;
+$cover-image-width: 1000px;
```
question for my curious mind: why do we make this width static instead of variable (i.e. a percentage of the viewport or something similar). How do we come up with this value, is it linked to another value?

ludwiczakpawel replied:
It's the size of the cover image that someone came up with, I don't know what was the reasoning for these number, maybe @benhalpern will know.

Even though we do define this width & height here, I make sure that wherever we display cover image it is actually responsive so it doesn't matter what size of the container is.

Also, setting exact px size will help users create artworks: so you can simply tell them "_create a cover photo with dimensions at least 1000px x 420px_" instead of "_create a cover photo with size ratio 50:21 width to height_" :D

Ridhwana commented on app/assets/stylesheets/views/article-form.scss:
```diff
@@ -0,0 +1,186 @@
+    }
+  }
+
+  &__tagsoptionrulesbutton {
```
another question for my curious mind, do we not use dashes or something similar with our crayons classnames, as a delimiter for words

ludwiczakpawel replied:
We do - this particular class though is not coming from crayons component BUT from one of our Preact components (but not crayons component): https://github.com/thepracticaldev/dev.to/blob/master/app/javascript/shared/components/tags.jsx - eventually we will probably refactor that to be crayonsified but that's just out of scope for this PR.

Ridhwana commented on app/assets/stylesheets/views/article.scss:
```diff
@@ -0,0 +1,364 @@
+      word-break: break-all;
+    }
+
+    kbd {
```
what is kbd?

ludwiczakpawel replied:
html element for representing keys on keyboard. literally :) https://www.w3schools.com/tags/tag_kbd.asp

Ridhwana commented on app/assets/stylesheets/views/article.scss:
```diff
@@ -0,0 +1,364 @@
+      }
+    }
+
+    h1,
```
Normally I thought that headings, tables, p styles etc. all  get defined at a high level app wide and we only override  a few if we want the style to change for that component. Do we approach it differently with our styling?

ludwiczakpawel replied:
That's usually true for blogs and other content heavy places that don't really contain other types of information and data...

DEV, even though is content heavy (articles) it also has pretty complex structure and data schema so we often use headings, paragraphs etc in many various ways - not only as representation of text-based content within post.

I think it's more common to actually "reset" styling for all kinds of elements globally and then redefine the styles for certain areas, like in article view. So we can keep using these elements in other places without inheriting original styling.

Ridhwana commented:
@ludwiczakpawel as per our chat, some points on this PR:
1. There's a lot of weird behaviour with the image uploader section disappearing and appearing, it will most likely be tricky to get 100% correct so for now we will make it static, at least to get this version out and then we can refine the behaviour as a second pass in another PR.
2. If I try and upload an image and then cancel it gives an error CarrierWave::IntegrityError. I'll double check this.
3. I can replicate the  `Whoops, something went wrong: main_image: is not a valid URL` on the master branch too, may have to do with local config or something, haven't looked into it.

ludwiczakpawel commented:
> let me know if I can be helpful in talking through anything on the TODOs

@benhalpern I think you could definitely help with tests - this is not one of my strengths.

ludwiczakpawel commented on app/javascript/article-form/articleForm.jsx:
```diff
@@ -331,185 +262,94 @@ export default class ArticleForm extends Component {
+        />
+
+        <div className="crayons-article-form__main">
+          {!previewShowing && (
```
good suggestions, i've improved it.

benhalpern commented:
Yeah @ludwiczakpawel I'll take a look

benhalpern commented:
So the failing tests mostly reflect expectations about what was on the page which are no longer there.

Going through and fixing it all at the end (or now) should be fairly straightforward...

<img width="545" alt="Screen Shot 2020-05-05 at 3 40 56 PM" src="https://user-images.githubusercontent.com/3102842/81108515-d11a3780-8ee6-11ea-85ae-3b5827eae20d.png">

There are still a number of glitches or slight UX issues with the current state of this. Let me know if/when this is in a good place for discussion of all the little things—We sometimes ship and then squash small issues after the fact, but I think we should adopt a fairly perfectionist attitude with this one given how critical it is in the user flow.

ludwiczakpawel commented:
> a number of glitches or slight UX issues with the current state of this. Let me know if/when this is in a good place for discussion of all the little things

@benhalpern you can just list everything here or slack or you can call me :D 100% agree about making it as perfect as possible before we ship!

benhalpern commented:
Alright, here are the issues I'm facing...

### Mobile still a little messed up

<img width="451" alt="Screen Shot 2020-05-07 at 8 53 58 PM" src="https://user-images.githubusercontent.com/3102842/81358781-724eec80-90a5-11ea-8235-2f1121c5dbc4.png">

### This one link takes you _away_ from the page, but the others open up the modal. We could maybe use a visual cue for a link that could take you away vs one that opens up a modal. (We could also do away with this particular phrase and cross that bridge later)

<img width="445" alt="Screen Shot 2020-05-07 at 8 51 22 PM" src="https://user-images.githubusercontent.com/3102842/81358782-72e78300-90a5-11ea-93ed-e27d4e631a5a.png">

### Tags should resemble final version (with #)

<img width="993" alt="Screen Shot 2020-05-07 at 8 50 11 PM" src="https://user-images.githubusercontent.com/3102842/81358784-73801980-90a5-11ea-945b-194104e0f757.png">

### Is this white space at the top of v1 intentional?

<img width="889" alt="Screen Shot 2020-05-07 at 8 47 54 PM" src="https://user-images.githubusercontent.com/3102842/81358788-74b14680-90a5-11ea-8414-511491f14156.png">

### Since all the settings go in the frontmatter with v1 we don't need the settings button

<img width="887" alt="Screen Shot 2020-05-07 at 8 47 28 PM" src="https://user-images.githubusercontent.com/3102842/81358789-74b14680-90a5-11ea-82df-5b7e801fc699.png">

### Clicking lower down in the white area does not put focus on the editor as I would expect it to.

<img width="948" alt="Screen Shot 2020-05-07 at 8 46 52 PM" src="https://user-images.githubusercontent.com/3102842/81358790-74b14680-90a5-11ea-8a60-a882e39664e0.png">

(Also no "current state" in edit/preview but that's in the issue checklist so I figure you're on that)

ludwiczakpawel commented:
thanks @benhalpern - most of these issues are already fixed.

Ridhwana commented:
**Update and Notes from Ridhwana:**
- Rspec tests are fixed. 
- Busy with snapshot testing for the js 
- Will tackle code climate errors thereafter

ludwiczakpawel commented:
_Carrierwave Integrity Error Check_ has been merged.

Ridhwana commented:
@ludwiczakpawel we're missing clear instructions on how to handle this on the new design: <img width="1636" alt="Screenshot 2020-05-13 at 19 47 31" src="https://user-images.githubusercontent.com/2786819/81846770-e6c8d600-9552-11ea-95a7-83ad743112a6.png">

ludwiczakpawel commented:
@Ridhwana I'm no sure if this is needed.. but i've added it inline anyway as a commented out frontmatter line: https://github.com/thepracticaldev/dev.to/pull/7524/commits/0c6d490db51f81dbbbfa1df757e6604788735166 do you think it's safe and makes sense? cc @benhalpern 

other than that I think this PR is ready on my side.. Is there anything else left on your plate?

ludwiczakpawel commented on app/javascript/article-form/articleForm.jsx:
```diff
@@ -331,185 +262,94 @@ export default class ArticleForm extends Component {
-        className={`articleform__form articleform__form--${version}`}
+        id="article-form"
+        className="crayons-article-form"
         onSubmit={this.onSubmit}
```
Good catch... Not sure if we should remove it or not...?

Ridhwana replied:
Hmmm Let me double check the effects and a trace through and will remove it if not needed 👍

Ridhwana commented:
Update: Tests almost complete. Did some cleanups as well. There were a couple of code climate issues that didn't really make sense, I need to head out but I'll send out a PR and link it here to merge into this branch.

Ridhwana commented:
@ludwiczakpawel the tests PR is ready https://github.com/ludwiczakpawel/dev.to/pull/4. I have a couple of thoughts/suggestions that I wanted to highlight below, some of which can be addressed in a subsequent PR:  

- article_body_markdown is underscored whereas the other help keys are dasherized (this can be amended in a separate PR though, as its not a dealbreaker and the PR is pretty big already)
- What do you think about changing Cover to CoverUploader to be more verbose, liek ImageUploader?
- In the cover image uploader, it is confusing when we call it cover_image vs main_image  - for example the key in frontmatter is cover_image but then the error comes back as `main_image is not a valid URL`. This is a bigger issue though not related to your PR that we should address in a different PR 
- Perhaps we don’t need to nest the Toolbar with the image uploader but we can just put the image uploader in the body? I know the thought here was for the disappearing of the uploader based on where you focus, but perhaps we can add the extra depth once we actually add that feature.

ludwiczakpawel commented:
> article_body_markdown is underscored whereas the other help keys are dasherized (this can be amended in a separate PR though, as its not a dealbreaker and the PR is pretty big already)

gotcha, yea lets do it in a separate PRs

> What do you think about changing Cover to CoverUploader to be more verbose, liek ImageUploader?

not 100% sold on it. ImageUploader is literally only image uploader. The Cover component is not only the uploader but also kinda a preview... But it's 2 out of 10 on my [sliding scale of giving a ****](https://capwatkins.com/blog/the-sliding-scale-of-giving-a-fuck). 

> Perhaps we don’t need to nest the Toolbar with the image uploader but we can just put the image uploader in the body? I know the thought here was for the disappearing of the uploader based on where you focus, but perhaps we can add the extra depth once we actually add that feature.

I'm not sure if I understand this. Can you elaborate?

ludwiczakpawel commented:
@Ridhwana would you be able to look into merging latest master into this branch? i think there were some changes in lots of places that we have touched on and I don't trust myself enough to be able to fix em properly...

ludwiczakpawel commented:
@benhalpern wanna give it for a final spin before we ship this fella?

benhalpern commented on app/views/articles/_v2_form.html.erb:
```diff
@@ -1,60 +1,93 @@
   <%== RunkitTag.script %>
 </script>
 
 <style>
```
This whole `style` area can be moved to the very top of the file since this is style that should be in place _as_ the body renders, not necessarily after (like when before it was just the footer).

So it can go before `<div id="article-form"` at the top.

ludwiczakpawel replied:
good suggestion ✅

benhalpern commented on app/javascript/crayons/Modal/Modal.jsx:
```diff
@@ -17,74 +17,66 @@ function getAdditionalClassNames({ size, className }) {
+            />
+          </div>
+        )}
+        <div className="crayons-modal__box__body">{children}</div>
```
There's an awkward side-scroll in this box since we have some contents which can be slightly wider than 100% I think. `overflow-x: hidden` would be a quick fix which would work for now if we don't want to immediately rewrite some of the inner functionality.

ludwiczakpawel replied:
i'll look into this.

benhalpern commented:
I feel like these tips should maybe fade in, maybe rise up slightly? Anyway, I think some kind of animation to make these a little less sudden would be nice.

<img width="416" alt="Screen Shot 2020-05-19 at 4 57 48 PM" src="https://user-images.githubusercontent.com/3102842/82377587-e1a3d500-99f1-11ea-8a9a-100c2585d6da.png">

joshpuetz commented:
If ever there was a PR that needed Percy.io, this is it!

benhalpern commented:
This is sooooo beautiful. I think it's pretty much good to go, but I added a few comments on small things which could change.

We're also going to do a quick workshop on the sidebar tip copy which shouldn't change functionality.

So make last mile changes you feel are appropriate and we'll integrate final-final copy and get this shipped tomorrow as far as I can see 🔥

benhalpern commented on app/javascript/article-form/components/TagsField.jsx:
```diff
@@ -0,0 +1,26 @@
+        onInput={onInput}
+        onFocus={switchHelpContext}
+        classPrefix="crayons-article-form"
+        fieldClassName="crayons-textfield crayons-textfield--ghost ff-accent fs-s s:fs-base"
```
This is `14px` in narrow screen mode which [will cause zoom on iOS](https://stackoverflow.com/questions/2989263/disable-auto-zoom-in-input-text-tag-safari-on-iphone) which is a weird experience. Not sure about equivalent on Android.... But I think all inputs will need to be minimum 16px in all scenarios to work beautifully on touch devices.

ludwiczakpawel replied:
good catch, should be fixed.

rhymes commented on app/services/articles/builder.rb:
```diff
@@ -91,7 +91,7 @@ def user_editor_v2
     def user_editor_v1
       Article.new(
-        body_markdown: "---\ntitle: \npublished: false\ndescription: \ntags: \n---\n\n",
+        body_markdown: "---\ntitle: \npublished: false\ndescription: \ntags: \n//cover_image: direct_url_to_image.jpg\n---\n\n",
```
I would change this to `// cover_image: https://direct_url_to_image.jpg` so it's clearer :D

ludwiczakpawel replied:
done.

rhymes commented on app/views/pages/_editor_frontmatter_help.html.erb:
```diff
@@ -0,0 +1,29 @@
+  <div class="crayons-article">
+    <div class="crayons-article__body">
+      <div class="crayons-notice crayons-notice--info">
+        <p><strong>We have two editor versions</strong>. If you'd prefer to edit title and tags etc. as separate fields, switch to the "rich + markdown" option in <a href="/settings/ux">/settings/ux</a>. Otherwise, continue...</p>
```
let's not harcode the path here, you can use `<%= user_settings_path(tab: :ux) %>`. I would also put "Settings -> UX" as the label

ludwiczakpawel replied:
good to know! thanks and fixed!

ludwiczakpawel commented:
@rhymes great, review, thank you!

> we should probably find a way to surface the list of supported HTML tags, it's currently hardcoded in the code but in a following PR we could display it there, what do you think?

+1

> markdown modal has horizontal scrolling

fixed.

> the liquid tag is missing a space and should probably have bolder headers

space is fixed. re bold: this modal content should actually mimic post styling which doesn't have bolded headlines BUT they are sans-serif (even if paragraphs are serifs).. so font face should be fixed, font weight probably not now :)

> it's weird how in the liquid modal some examples have black bands and others are inline, i think we should have code blocks for each one):

agree! but help content is coming from current help pages, so i'll keep it as is for now and address that in a separate PR. there're probably more low-hanging fruits in that Help section.

> tab selection of a tag is broken but i think it's broken in production as well, you can only select with the enter key or the mouse

correct. as you pointed out - that's also an issue in current Editor. i don't want to address it in this PR because tags picker is being used in other places and my PR is already big enough :D 

> there is no visual indication that the upload is happening, we might want to address this in a PR right after as it's a bit weird to not have visual feedback while waiting for the upload to happen

we have an [image uploader designed](https://www.figma.com/file/PIPeiiVSADmcpCbiZTdJm9/Write-a-Post?node-id=3%3A667) but were not able to ship it yet.. so this will have to come in future. wanna work on it? :)))

> I expected the gear button modal to open above and aligned with the left side of button, not there

fixed.

>  gear button may need a title attribute

added.

> I wonder if instead of "Edit|Preview" it shouldn't say "Write|Preview"

i'm not sold on either :P but because "write" is longer, i'll stick to "Edit" so for the smallest breakpoints we save some pixels (very bad argument though :D) 

---

i'll be also looking into accessibility stuff! thank you so much!!

ludwiczakpawel commented:
@benhalpern I've addressed all of the thing (still waiting for copy updates) but I'm having some issues with this:

> some kind of animation to make these a little less sudden would be nice.

So I've added a subtle animation and it works but not always... Sometimes when I switch between fields the animation is there and sometimes it's not... Any ideas? https://github.com/thepracticaldev/dev.to/pull/7524/commits/94a0b6ea094fab5760ee3bc8e44bc4ddf3fcea30

rhymes commented:
@ludwiczakpawel thanks for the changes! I can take a look at some of the a11y issues to see if there are low hanging fruits and then we review it again after deployment. What do you think?

rhymes commented:
@ludwiczakpawel 

Another thing I noticed: we have multi image uploads!

Should the text box expand when there's more than one image?

![Screenshot_2020-05-20 New Post - DEV(local) Community 👩‍💻👨‍💻(2)](https://user-images.githubusercontent.com/146201/82452402-7e787800-9aaf-11ea-8dd4-f9b36f7a1b3d.png)

rhymes commented:
@nickytonline @ludwiczakpawel @Ridhwana I added some a11y fixes in https://github.com/thepracticaldev/dev.to/pull/7524/commits/43ad76045bdf2b61086e8aae84e1f59c785386bb

I wish we could use `inline_svg_tag` or `inline_svg_pack_tag` from https://github.com/jamesmartin/inline_svg instead of having to copy paste the SVG inside the Preact component :D

Ridhwana commented:
Just a note the code climate Issues here seems superfluous. I've outlined the reasons why in this thread https://dev-internal.slack.com/archives/C7GE84DEJ/p1589896547188700

ludwiczakpawel commented:
> Should the text box expand when there's more than one image?

As mentioned in previous comment - we're going to improve the image uploader feature (as shown in the design... hey! wanna work on that? :D) so I don't think I'll be doing any other improvements here for now.

thank you for helping out with all the a11y things!! i really appreciate it <3

nickytonline commented on app/assets/stylesheets/components/tabs.scss:
```diff
@@ -3,7 +3,7 @@
   // Setup
   --border-width: 3px;
-  --tab-vertical-padding: var(--su-3);
+  --tab-vertical-padding: var(--su-2);
```
Same goes for tabs. This could be another PR.

ludwiczakpawel replied:
Similar to what I mentioned re Modals: these Tabs, as frontend component, are quite new addition that isn't really used anywhere except Notifications (which I've introduced last week) so I'm quite positive this change will not have negative consequences in other places.

nickytonline replied:
I realize it's not used yet. I'm just looking for ways to break up the PR. 😃

ludwiczakpawel replied:
Yea I know. And as I mentioned in one of our calls - big PRs are bad :). But that one line won't really change perception of this PR :D:D:D:D

nickytonline commented on app/assets/stylesheets/typography.scss:
```diff
@@ -3,7 +3,8 @@
 // Alternate base fonts
 body.sans-serif-article-body {
-  .body {
+  .body,
```
These changes seem related to a post and not the editor. If so this could be another PR.

ludwiczakpawel replied:
This is tightly related to this PR since it's for post preview in the editor.

nickytonline commented on app/javascript/article-form/components/Actions.jsx:
```diff
@@ -0,0 +1,134 @@
+import { Button } from '@crayons';
+import { Options } from './Options';
+
+export class Actions extends Component {
```
`Actions` is too generic a name. Consider naming it something like `EditorActions` or `EditorToolbar`.

ludwiczakpawel replied:
Renamed to `EditorActions`.

nickytonline commented on app/javascript/article-form/components/Body.jsx:
```diff
@@ -0,0 +1,38 @@
+import Textarea from 'preact-textarea-autosize';
+import { Toolbar } from './Toolbar';
+
+export const Body = ({
```
This class name is too generic as well, and from a DOM standpoint, it looks confusing. `<Body />` vs. `<body />`. Consider a more specific name.

ludwiczakpawel replied:
Renamed to `EditorBody`.

nickytonline commented on app/javascript/article-form/components/Cover.jsx:
```diff
@@ -0,0 +1,101 @@
+import { generateMainImage } from '../actions';
+import { validateFileInputs } from '../../packs/validateFileInputs';
+
+export class Cover extends Component {
```
The name of this class is not clear to me. Consider a more specific name, e.g. `<PostCoverImage />` or `<ArticleCoverImage />`

ludwiczakpawel replied:
Renamed to `ArticleCoverImage`.

nickytonline commented on app/javascript/article-form/components/Cover.jsx:
```diff
@@ -0,0 +1,101 @@
+  render() {
+    const { mainImage } = this.props;
+    const { uploadError, uploadErrorMessage } = this.state;
+    const uploadLabel = mainImage ? 'Change' : 'Add a cover';
```
```suggestion
    const uploadLabel = mainImage ? 'Change' : 'Add a cover image';
```

nickytonline commented on app/javascript/article-form/components/Errors.jsx:
```diff
@@ -0,0 +1,27 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+
+export const Errors = ({ errorsList }) => {
+  return (
+    <div className="crayons-notice crayons-notice--danger mb-6">
+      <h3 className="fs-l mb-2 fw-bold">Whoops, something went wrong:</h3>
+      <ul className="list-disc pl-6">
+        {Object.keys(errorsList).map((key) => {
+          return (
+            <li>
+              {key}
+              {`: `}
+              {errorsList[key]}
+            </li>
+          );
+        })}
+      </ul>
+    </div>
+  );
+};
+
+Errors.propTypes = {
+  errorsList: PropTypes.objectOf(PropTypes.string).isRequired,
+};
+
+Errors.displayName = 'Errors';
```
```suggestion
export const ErrorList = ({ errors }) => {
  return (
    <div className="crayons-notice crayons-notice--danger mb-6">
      <h3 className="fs-l mb-2 fw-bold">Whoops, something went wrong:</h3>
      <ul className="list-disc pl-6">
        {Object.keys(errors).map((key) => {
          return (
            <li>
              {key}
              {`: `}
              {errors[key]}
            </li>
          );
        })}
      </ul>
    </div>
  );
};

ErrorList.propTypes = {
  errors: PropTypes.objectOf(PropTypes.string).isRequired,
};

ErrorList.displayName = 'Errors';
```

ludwiczakpawel commented on app/assets/stylesheets/base/layout.scss:
```diff
@@ -1,9 +1,29 @@
 @import '../config/import';
 
 :root {
+  --layout-site-width: var(--site-width);
```
these are in fact for the editor view, and will later on be used in more places too.

ludwiczakpawel commented on app/assets/stylesheets/components/modals.scss:
```diff
@@ -2,7 +2,7 @@
 .crayons-modal {
   --modal-max-width: 768px;
-  --modal-body-padding: var(--su-2);
+  --modal-body-padding: var(--su-3);
```
Since this part of the modal itself is not being used anywhere else in DEV, I've decided to include it in this PR since it is actually needed there too: it addresses overflow issue @benhalpern has mentioned in earlier comments.

nickytonline replied:
Isn't the modal being used in onboarding?

ludwiczakpawel replied:
Parts of it, yes. But I have tested it the onboarding after this change ✅

nickytonline commented:
I love emojis, but the eyes emojis just seems out of place here. I get that it means preview, but it looks a little odd in between the cover image and content.

![image](https://user-images.githubusercontent.com/833231/82482957-9c40df80-9aa5-11ea-84a3-f2ecb3ab35e3.png)

nickytonline commented:
Nice touch with the help info for each section of the post in the editor.

![image](https://user-images.githubusercontent.com/833231/82483113-e033e480-9aa5-11ea-86f1-970ef71efd9c.png)

benhalpern commented:
> I love emojis, but the eyes emojis just seems out of place here. I get that it means preview, but it looks a little odd in between the cover image and content.

Oh yeah I agree with Nick there. I find it kind of confusing.

ludwiczakpawel commented:
@nickytonline @benhalpern I've addressed the issues you mentioned.

re help on mobile: I'm fully aware of that - I'm working on a mobile pattern to show help without disturbing user while in the editor, but it's not there yet so I'd address it afterwards. I deprioritized it since writing posts on mobile is not (yet) a crazy popular thing to do on DEV, but definitely something we will want to make right later on.

Ridhwana commented:
Some tests broke after the latest changes :( Not at my laptop at the moment to fix them, I can look at them in the morning if the plan is to merge tomorrow US time.

ludwiczakpawel commented:
good catch. i think i’ve fixed it.

vaidehijoshi commented:
@ludwiczakpawel just a heads up that we now have Percy in production, so if this PR was made from the main repo (rather than a fork), you'd see at least some of your changes reflected in the Percy snapshots!

nickytonline commented on app/javascript/article-form/components/Preview.jsx:
```diff
@@ -41,76 +30,53 @@ function titleArea(previewResponse, version, articleState) {
-      />
-    );
-  }
+  const previewTitle = previewResponse.title || articleState.title;
```
Any reason the `|| '';` is being omitted?

ludwiczakpawel replied:
not really 👍. i brought it back.

nickytonline commented:
I've found a bug that is super easy to reproduce. Before adding a title, press the Publish button. The error message about a title can't be blank appears as expected, but the Publish button changes to a Save Changes button and there is no longer a Save Draft button.

![May-21-2020 14-23-50](https://user-images.githubusercontent.com/833231/82592633-36208f00-9b6f-11ea-8e67-7125dacb45eb.gif)

This PR is quite large and is going to have to undergo more thorough testing before we put it in the wild. There needs to be more tests added along with manual testing and I still do not have a local setup for cover image uploading. Are you set up to test that @benhalpern?

In the meantime, I will fix this bug.

ludwiczakpawel commented on app/javascript/article-form/components/Errors.jsx:
```diff
@@ -0,0 +1,27 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+
+export const ErrorList = ({ errors }) => {
```
Pull the latest, I think it should be renamed

ludwiczakpawel commented:
> I've found a bug that is super easy to reproduce. Before adding a title, press the Publish button. The error message about a title can't be blank appears as expected, but the Publish button changes to a Save Changes button and there is no longer a Save Draft button.

Definitely a bug, but also happens on prod now so it's not PR related so I think it could be fixed in a separate PR (if this PR isn't big enough :D)

nickytonline commented:
> > I've found a bug that is super easy to reproduce. Before adding a title, press the Publish button. The error message about a title can't be blank appears as expected, but the Publish button changes to a Save Changes button and there is no longer a Save Draft button.
> 
> Definitely a bug, but also happens on prod now so it's not PR related so I think it could be fixed in a separate PR (if this PR isn't big enough :D)

Ahh, didn't realize it was a bug in production. I'll fix it in this PR anyways since this will be the new editor for v2.

nickytonline commented:
The publish issue mentioned above has been fixed in the last commit.

![May-21-2020 17-27-58](https://user-images.githubusercontent.com/833231/82608371-8a843880-9b88-11ea-8e56-473339a9f975.gif)

nickytonline commented:
Things appear to be working after doing some refactoring and fixing some tests. @benhalpern and @ludwiczakpawel, let's take it for one last spin on Tuesday once everyone is back and then this should be good to go.

ludwiczakpawel commented:
@nickytonline sweet, thanks for jumping in!

nickytonline commented:
@ludwiczakpawel, you'll need to merge latest from master and then run `yarn test -u` to update the snapshots mentioned above.

ludwiczakpawel commented:
@Ridhwana I tried to merge and fix conflicts but im not sure if i didn't break something.. would you mind looking at last commits again? not sure what's happening there :x

ludwiczakpawel commented:
> @Ridhwana I tried to merge and fix conflicts but im not sure if i didn't break something.. would you mind looking at last commits again? not sure what's happening there :x

..or @nickytonline ? :)

nickytonline commented:
> > @Ridhwana I tried to merge and fix conflicts but im not sure if i didn't break something.. would you mind looking at last commits again? not sure what's happening there :x
> 
> ..or @nickytonline ? :)

I'll take a peek @ludwiczakpawel .

nickytonline commented:
Should be all good now @ludwiczakpawel.

ludwiczakpawel commented:
@mstruve @nickytonline oh wow something went off..

nickytonline commented:
@mstruve @ludwiczakpawel oh snap, I merged latest from master and 402 file changes definitely does not look good. 🙃 I'll see what's up

nickytonline commented:
Should be all good now @mstruve. I screwed up on a git revert of a merge commit twice! YOLO

ludwiczakpawel commented:
@benhalpern @Ridhwana it should be ready whenever we are.. not sure about codeclimate issues?

Ridhwana commented:
@ludwiczakpawel I think this should be merged and deployed US time, in the case any issues arise. The code climate issues can be bypassed by an admin as they are not relevant.

nickytonline commented:
The issue that @benhalpern reported for the tags section has been fixed. I'd say take this for one last spin and then let's get this sucker merged.

nickytonline commented:
I'd say merge it today @ludwiczakpawel. There's codeclimate issues that I can't override, so @benhalpern or @maestromac, can you merge this?
